Test Pipeline YAML context menu and verify read only

manosnoam opened the pull request:
- Test that the context menu (right click) of a Pipeline YAML has a command menu, and responds to user typing "Fold all" yaml tags.
- Verify YAML text is readonly (Bug https://github.com/opendatahub-io/odh-dashboard/issues/1689)

manosnoam commented:
This new test has been verified:
![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/87efaf82-cdc1-4891-8fd4-86f1a0eb9a8b)

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/Pipelines.resource:
```diff
@@ -112,6 +112,24 @@ Click Action From Actions Menu
+    [Arguments]    ${pipeline_name}
+    ${orig wait}=    SeleniumLibrary.Set Selenium Implicit Wait    10s
+    Wait Until Element is Visible     //button[contains(text(), "Data Science Pipelines")]          timeout=30
+    Click Button      Data Science Pipelines
+    Click Link      Pipelines
```
you can use this keyword to move in the desired section`Menu.Navigate To Page`

manosnoam replied:
Thanks - fixed it.

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/Pipelines.resource:
```diff
@@ -112,6 +112,24 @@ Click Action From Actions Menu
+    Wait Until Element is Visible     //a[text()="${pipeline_name}"]          timeout=30s
+    Click Link      ${pipeline_name}
+    Click Element     //button[@aria-label="Pipeline YAML Tab"]
+    Open Context Menu    //div[contains(@class, 'lines-content')]
+    Click Element     //button[@aria-label="Pipeline YAML Tab"]
```
by looking at the GH issue, i understood that there are 2 different scenario:
1. the pipeline YAML editor should be read only
2. the Command Palette should not allow to run commands. I guess this is the "Context Menu" this keyword is referring.

Is my understanding correct? If yes, I would expect this "Open Context Menu" to be placed either in a separated keyword or after checking the Pipeline YAML editor is read-only

manosnoam replied:
The commands menu should allow to run commands, and this is tested just by a single line of code:
`Press Keys    //div[contains(@class, 'lines-content')]    F1+fold+SPACE+all+ENTER`

If I move this line to a new keyword, we will have to rerun the prerequisites to reach a pipeline yaml in that kw, which will also take more execution time, and duplicating code.

In case we decide to add a complete new test, for example to test all the available command options - if we see it is high priority use case (I don't think it is though) - we can consider to add a new test with more keywords.

For current use case I believe keeping the readonly validation + context check in same keyword is fine.

bdattoma replied:
My concern is that we don't check that the YAML is read-only by trying editing it without using the context-menu.

For example, you could have the pre-requisites in the test. Then you have a kw to test the context menu and you have a kw to try editing the YAML from the Editor (without context menu).

manosnoam replied:
In the context menu the "edit" commands do not exists. That's why I need to send keyboard types to check if edit is not allowed.

bdattoma replied:
yeah I'm suggesting to try editing without leveraging on context menu

manosnoam replied:
Yes, that is what the logic does: 
1) Open Context Menu with mouse to verify it is available.
2) Close the menu by clicking the YAML again
3) Open the context menu with F11 and send command to fold all.
4) Try to edit the YAML without the context menu, by sending keyboard types

bdattoma replied:
could you make more clear it in the code? The kw is called "Pipeline Context Menu Should Be Readonly" but actually mixes the context menu with Editor check. I think also for future maintanance it would be better to have clear steps in the test.
You could move the pre-requisites steps outside the kw and run them once. But then with separated keywords you check the editor is read-only in any available scenarios (context menu vs direct typing in the Editor)

manosnoam replied:
Separated to 2 keywords:

![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/d4c3a366-20cf-4f80-9630-962ea254b2c0)

manosnoam replied:
nope, it's testing one of the available commands ("fold all"), and verifies both mouse reaction and keyboard

bdattoma commented on ods_ci/tests/Resources/Page/ODH/ODHDashboard/ODHDataScienceProject/Pipelines.resource:
```diff
@@ -112,6 +112,24 @@ Click Action From Actions Menu
+    Click Element     //button[@aria-label="Pipeline YAML Tab"]
+    Press Keys    //div[contains(@class, 'lines-content')]    F1+fold+SPACE+all+ENTER
+    Press Keys    //div[contains(@class, 'lines-content')]    cannot_enter_read_only
+    Wait Until Element Is Visible    //div[@class="pf-c-code-editor pf-m-read-only odh-dashboard__code-editor"]    timeout=10
```
shouldn't this be placed after `Click Element //button[@aria-label="Pipeline YAML Tab"]` in order to wait for the Editor to be rendered?

manosnoam replied:
Not sure I understand your question... 
The implicit wait I added removes the requirement to add explicit wait for elements.

bdattoma replied:
hm so why do you need to wait here? is the page or the element reloaded?

manosnoam replied:
Here I wait explicitly - To see that the "Read Only" alert becomes visible (it is hidden by default). 
See it in the bottom of screenshot:
https://github.com/red-hat-data-services/ods-ci/pull/995#issuecomment-1784051461

bdattoma commented on ods_ci/tests/Tests/400__ods_dashboard/430__data_science_pipelines/434__data-science-pipelines-ui.robot:
```diff
@@ -105,11 +109,8 @@ Verify Pipeline Metadata Pods Are Not Deployed When Running Pipelines
     Set Library Search Order    SeleniumLibrary
+    # TODO: Install Pipeline only if it does not already installed
     Install Red Hat OpenShift Pipelines
-    ${prj_title}=    Set Variable    ${PRJ_BASE_TITLE}-${TEST_USER_3.USERNAME}
-    ${iris_pipeline_name}=    Set Variable    ${PIPELINE_TEST_BASENAME}-${TEST_USER_3.USERNAME}
-    Set Suite Variable    ${PRJ_TITLE}    ${prj_title}
-    Set Suite Variable    ${PIPELINE_TEST_NAME}    ${iris_pipeline_name}
```
why removing these lines?

manosnoam replied:
Since these are widely used variables, I moved them under *** Variables *** section:
```
${PRJ_BASE_TITLE}=   dsp
${PRJ_DESCRIPTION}=   ${PRJ_BASE_TITLE} is a test project for validating DS Pipelines feature
```
https://github.com/red-hat-data-services/ods-ci/pull/995/files#diff-0e3e8f6733ca76331b5ae98a47739e91267cb03bd98a02b4d57615fe8ea4480aR16-R17

bdattoma replied:
ah, okay :)

manosnoam commented:
Changed to use Menu navigation kw, rebased and tested:

![image](https://github.com/red-hat-data-services/ods-ci/assets/9913945/dba1eca3-d58b-4f78-88b9-e706d82aa8aa)
